prototype2/client/src/engine/control/console.rs

```diff
@@ -0,0 +1,272 @@
+      CharEvent::Action(CharAction::Backspace) => {
+        if self.command_buffer.len() > 0 {
+          self.command_buffer.remove(self.cursor.saturating_sub(1));
+          self.cursor = self.cursor.saturating_sub(1);
+        }
+      },
+      CharEvent::Action(CharAction::Delete) => {
+        if self.command_buffer.len() > self.cursor {
+          self.command_buffer.remove(self.cursor);
+        }
+      },
+      CharEvent::Action(CharAction::Return) => {
+        let mut buffer = String::new();
+        mem::swap(&mut buffer, &mut self.command_buffer);
+        self.cursor = 0;
+        executed_commands.push(ExecutedCommand(buffer));
+      },
+      CharEvent::Invalid => {},
+    };
+  }
+}
+
+impl specs::System<engine::Delta> for System {
+  fn run(&mut self, arg: specs::RunArg, _: engine::Delta) {
+    use glutin::VirtualKeyCode::{Capital, LShift, RShift};
+    use glutin::ElementState;
+    use itertools::Itertools;
+
+    let (mut console_events,
+         mut console_command_buffer,
+         mut console_cursor,
+         mut executed_commands) = arg.fetch(|w| {
+      (w.write_resource::<Vec<ConsoleEvent>>(),
+       w.write_resource::<CommandBuffer>(),
+       w.write_resource::<CommandCursor>(),
+       w.write_resource::<Vec<ExecutedCommand>>())
+    });
+
+    console_events.drain(..).foreach(|event| {
+      match event.get_input() {
+        &KeyboardInput(ElementState::Pressed, _, Some(LShift)) => {
+          self.lshifted = true;
+        },
+        &KeyboardInput(ElementState::Released, _, Some(LShift)) => {
+          self.lshifted = false;
+        },
+        &KeyboardInput(ElementState::Pressed, _, Some(RShift)) => {
+          self.rshifted = true;
+        },
+        &KeyboardInput(ElementState::Released, _, Some(RShift)) => {
+          self.rshifted = false;
+        },
+        &KeyboardInput(ElementState::Pressed, _, Some(Capital)) => {
+          self.capsed = !self.capsed;
+        },
+        &KeyboardInput(ElementState::Pressed, _, Some(ref key)) => {
+          let event = input_to_char_event(key.clone(), self.should_cap());
+          self.handle_char_event(event, &mut executed_commands);
+        },
+        &KeyboardInput(ElementState::Released, _, _) => {},
+        _ => {}, // Something thats not a keyboard input
+      }
+    });
+
+    // We're authoritative for the console command and the cursor -- any mutation
+    // gets overriden
+    *console_command_buffer = CommandBuffer(self.command_buffer.clone());
+    *console_cursor = CommandCursor(self.cursor);
+  }
+}
+
+pub enum CharMotion {
+  Left,
+  Right,
+  Down,
+  Up,
+  Home,
+  End,
+}
+
+pub enum CharAction {
+  Backspace,
+  Delete,
+  Return,
+}
+
+pub enum CharEvent {
+  Character(char),
+  Motion(CharMotion),
+  Action(CharAction),
+  Invalid,
+}
+
+fn input_to_char_event(code: VirtualKeyCode, shift: bool) -> CharEvent {
+  use glutin::VirtualKeyCode::*;
+
+  match (shift, code) {
```

Comment:
something about these mechanical mappings make me feel like they should live in their own file, some statically defined dictionary. just a thought